### ISSUE TITLE
[Bug] `is_reg` 写错 TOML 层级导致训练必崩，且被误诊为 bitsandbytes 8-bit 问题

### ISSUE THREAD

**Issue by YSBGQAQ** [l1934332574-maker/Kohya-LoRA-Tool#3]
# [Bug] 人物模式 + 正则数据集：`is_reg` 写错 TOML 层级导致训练必崩，且被误诊为 bitsandbytes 8-bit 问题

## 环境

- 版本：v0.9.28（Setup.exe 安装）
- 系统：Windows 10 Pro
- 显卡：NVIDIA 16GB（日志显示"显存 15.6GB 智能适配"）
- 训练配置：Anima 底模 + 👤 人物角色LoRA模式 + **正则数据集（reg_dir）** + WD14 自动打标

## 复现步骤

1. 新建项目，选择人物角色LoRA模式，底模选 Anima（SD15/SDXL 等走第一引擎的架构同样受影响）
2. 填写 Trigger，**并填写正则数据集文件夹**（本问题的关键条件）
3. 点击「🚀 一键开始训练」

预处理正常完成，训练启动后立刻崩溃，退出码 1。

## 报错关键日志

```
voluptuous.error.MultipleInvalid: extra keys not allowed @ data['datasets'][1]['is_reg']
```

随后 GUI 给出的诊断是**误诊**：

```
[训练] ⚠ 检测到训练失败与 bitsandbytes 8-bit 优化器(AdamW8bit)相关（命中：8bit）
[训练] 原因：Windows + CUDA 下 bitsandbytes 的 8-bit CUDA 内核缺失/不兼容，训练到 optimizer.step() 阶段崩溃。
```

实际上 AdamW8bit 优化器预检是通过的（日志前一行：`[优化器] AdamW8bit 预检通过（真实 step 成功）`），训练根本没跑到优化器阶段——是在加载 `dataset_config.toml` 时就被 voluptuous 拒绝了。

### 1. `is_reg` 被写在了数据集层级，而不是子集层级

`preprocess.py` 的 `write_dataset_config()` 在 `reg_dir` 非空时，把 `is_reg = true` 写进了 `[[datasets]]` 块：

```toml
[[datasets]]
resolution = 1024
batch_size = 1
is_reg = true          # ← 写错层级：这里不被 schema 允许
...

  [[datasets.subsets]]
  image_dir = "..."
  num_repeats = 1
```

而 sd-scripts 的 `ConfigSanitizer.DB_SUBSET_DISTINCT_SCHEMA` 定义中，`is_reg` 是**子集（subset）级**键，必须写在 `[[datasets.subsets]]` 内。因此任何「人物模式 + 正则数据集」的组合都 100% 触发本错误。

### 2. 打包版 exe 即使手动修了外部 preprocess.py 也会复发

`Kohya一键工具.py` 在启动训练前会 `import preprocess as _pp` 并再次调用 `write_dataset_config()` 重写 TOML。PyInstaller 已将这个 import 静态冻结进 exe 内部（`preprocess.py` 同时以 datas 外部文件 + PYZ 冻结模块两种形式存在），**冻结模块优先级高于磁盘上的外部 .py**。

结果：外部 preprocess.py 修好后，预处理子进程写出的配置是对的，但 GUI 在训练前又用内嵌的旧版模块把配置覆盖回坏格式——表现为"怎么修都会复发"，只有重新打包 exe 才能根治。

### 3. 失败诊断按裸关键字 `"8bit"` 匹配，误报 bitsandbytes

`_diagnose_optimizer_failure()` 的关键字列表含 `"8bit"`，且只过滤了 `$ ` 开头的命令行回显。但 `CalledProcessError` 会以 `Command '[...]'` 的形式**整行重印训练命令**（其中必含 `--optimizer_type=AdamW8bit`），未被过滤，导致任何退出码非 0 的失败（包括本次的配置校验错误、OOM 等）只要用了 AdamW8bit 就会被误诊为 bitsandbytes 问题，把用户引向完全错误的方向。

### 修复一：`preprocess.py` — `is_reg` 移到子集块内（根因）

```diff
     if reg_dir and os.path.isdir(reg_dir):
         reg_subsets = reg_subsets or [(reg_dir, 1)]
         lines += [
             "",
             "[[datasets]]",
             f"resolution = {resolution}",
             f"batch_size = {batch_size}",
-            "is_reg = true",
             "enable_bucket = true",
             "bucket_no_upscale = true",
             "bucket_reso_steps = 64",
             f"min_bucket_reso = {min_bucket}",
             f"max_bucket_reso = {max_bucket}",
         ]
         for _item in reg_subsets:
             _d = _item[0]
             _nr = _item[1]
             _abs = os.path.abspath(_d).replace("\\", "/")
             lines += [
                 "",
                 "  [[datasets.subsets]]",
                 f'  image_dir = "{_abs}"',
                 f"  num_repeats = {int(_nr)}",
+                "  is_reg = true",
             ]
```

### 修复二：`Kohya一键工具.py` — `_diagnose_optimizer_failure()` 修正误诊

```diff
 def _diagnose_optimizer_failure(opt_k, log_text, logf=print):
-    """训练失败后诊断：用了 AdamW8bit 且日志命中 bitsandbytes 8-bit 崩溃关键字时，给出明确提示。
+    """训练失败后诊断：优先识别数据集配置校验错误，其次判断 bitsandbytes 8-bit 崩溃。

-    opt_k 为空时自动从日志命令行解析 --optimizer_type=...。
-    返回 True=命中 bnb 8-bit 问题（调用方可在报错信息里追加建议）。
+    opt_k 为空时自动从日志命令行解析 --optimizer_type=...。
+    返回 True=已给出明确诊断（调用方可在报错信息里追加建议）。
     """
     if not log_text:
         return False
+    # 数据集配置校验失败（voluptuous schema 拒绝 dataset_config.toml）：与优化器无关，
+    # 必须优先报告真实原因，避免误导用户去改优化器
+    if "Invalid user config" in log_text or "extra keys not allowed" in log_text or "MultipleInvalid" in log_text:
+        m = re.search(r"extra keys not allowed\s*@\s*(\S+)", log_text)
+        where = f"（非法键位置：{m.group(1)}）" if m else ""
+        logf("[训练] ⚠ 检测到数据集配置 dataset_config.toml 校验失败" + where)
+        logf("[训练] 原因：TOML 中存在 sd-scripts 不认识的键，或键写错了层级"
+             "（例如 is_reg 必须写在 [[datasets.subsets]] 子集层级，不能写在 [[datasets]] 层级）。")
+        logf("[训练] 解决：重新执行数据预处理让工具重写配置，或手动修正 configs/dataset_config.toml 后重试。")
+        return True
     if not opt_k:
         m = re.search(r"(?:^|\s)--?optimizer_?type[=:\s]+([A-Za-z0-9_]+)", log_text, re.I)
         opt_k = m.group(1) if m else ""
     if str(opt_k).lower() not in ("adamw8bit", "adamw_8bit", "8bit"):
         return False
     keys = ("libbitsandbytes", "str2optimizer8bit_blockwise", "bitsandbytes",
-            "compiled without GPU support", "cuda setup", "8-bit optimizer", "8bit")
-    # 剔除 $ 开头的命令行（里面必含 --optimizer_type=AdamW8bit，会污染 "8bit" 关键字匹配），
-    # 只在实际输出里找 bnb 崩溃特征
-    low = "\n".join(l for l in log_text.splitlines() if not l.lstrip().startswith("$ ")).lower()
+            "compiled without GPU support", "cuda setup", "8-bit optimizer")
+    # 剔除命令行回显：$ 开头的命令行，以及 CalledProcessError 里 "Command '[...]'" 的整行重印，
+    # 两者都必含 --optimizer_type=AdamW8bit，会污染关键字匹配；只在实际输出里找 bnb 崩溃特征
+    low = "\n".join(
+        l for l in log_text.splitlines()
+        if not l.lstrip().startswith("$ ") and "command '[" not in l.lower()
+    ).lower()
```

## 验证方式

- 应用修复一后重新生成 TOML，用 sd-scripts 的 `ConfigSanitizer(True, True, False, True).sanitize_user_config()` 校验通过，`is_reg=True` 正确落在子集层级
- 端到端实测：用真实训练命令启动 Anima LoRA 训练，配置校验通过，正则数据集 8 张被正确加载（`num reg images: 8`），训练正常开跑（本机已确认跑到第 7/1290 步，loss 正常）
- 修复二实测：同一份失败日志现在正确报「数据集配置校验失败（非法键位置：datasets'][1]['is_reg']）」；真实 bitsandbytes 崩溃日志仍能正确命中；OOM 日志不再误报

## 补充建议

`preprocess.py` 目前以「datas 外部文件 + PYZ 冻结模块」双形式打包进 exe，两份副本容易不同步（本 bug 就是外部文件和内嵌模块行为不一致导致的排查困难）。建议二选一：

- 训练前重写 TOML 时改用 `subprocess` 调用外部 `preprocess.py`（与预处理步骤保持同一代码路径）；或
- 不把 `preprocess.py` 作为 datas 打包，统一从冻结模块读取（并在文档说明）

---

感谢开发这么好用的一键工具，整体引导和智能适配体验很棒！
